File: prolog/src/sprint.pl
:- module(sprint, [criar_sprint/1, listar_sprints_da_empresa/1, acessar_sprint/2, adicionar_tarefa_a_sprint/1, atribuir_tarefa/1]).

:- use_module(usuario).
:- use_module(tarefa).

:- dynamic sprint/6.

% Função para criar uma nova sprint
criar_sprint(Usuario) :-
    write('Digite o ID da Sprint:'), nl,
    read(Id),
    write('Digite o Nome da Sprint:'), nl,
    read(Nome),
    write('Digite a Duração da Sprint (em dias):'), nl,
    read(Duracao),
    usuario:usuario_empresa_id(Usuario, EmpresaId),
    assertz(sprint(Id, Nome, Duracao, [], usuario_id(Usuario), EmpresaId)),
    format('Sprint criada: ~w~n', [Id]).

% Função para listar as sprints da empresa do usuário
listar_sprints_da_empresa(Usuario) :-
    usuario_empresa_id(Usuario, EmpresaId),
    findall(Sprint, sprint(_, Sprint, _, _, _, EmpresaId), SprintsDaEmpresa),
    (   SprintsDaEmpresa = []
    ->  writeln('Não há sprints nesta empresa.')
    ;   write('Sprints da Empresa:'), nl,
        maplist(print_sprint, SprintsDaEmpresa),
        write('Digite o ID da sprint para visualizar suas tarefas (ou -1 para voltar, 0 para criar uma nova sprint):'), nl,
        read(Entrada),
        processar_entrada(Entrada, Usuario)  % Processa a entrada do usuário
    ).

% Função para processar a entrada do usuário
processar_entrada(-1, _) :- !.  % Voltar sem alterar nada
processar_entrada(0, Usuario) :- 
    criar_sprint(Usuario),
    listar_sprints_da_empresa(Usuario).  % Lista novamente após criar a sprint
processar_entrada(Entrada, Usuario) :- 
    (   sprint(Entrada, SprintEscolhida, _, _, _, _) -> 
        acessar_sprint(Usuario, SprintEscolhida)  % Acessar a sprint escolhida
    ;   writeln('Sprint não encontrada.'),
        listar_sprints_da_empresa(Usuario)  % Voltar a listar sprints
    ).

% Acessar a sprint e exibir suas informações
acessar_sprint(Usuario, Sprint) :-
    sprint(Sprint, Nome, _, _, _, _),
    format('Sprint Selecionada: ~w~n', [Nome]),
    write('Tarefas da Sprint:'), nl,

    % Filtrar as tarefas da sprint
    findall(TarefaId, tarefa_id(TarefaId, Sprint), TarefasDaSprint),
    
    (   TarefasDaSprint = []
    ->  writeln('Nenhuma tarefa nesta sprint.')
    ;   maplist(print_tarefa, TarefasDaSprint)
    ),
    
    % Exibir o menu
    write('Escolha uma opção:'), nl,
    write('1. Adicionar Tarefa à Sprint'), nl,
    write('2. Atribuir Tarefa a um Usuário'), nl,
    write('0. Voltar'), nl,
    read(Escolha),
    
    (   Escolha = 1 -> 
        adicionar_tarefa_a_sprint(Sprint),  % Adiciona tarefa à sprint
        acessar_sprint(Usuario, Sprint)  % Rechama após adicionar tarefa
    ;   Escolha = 2 -> 
        atribuir_tarefa(Usuario),  % Atribui tarefa a um usuário
        acessar_sprint(Usuario, Sprint)  % Rechama após atribuir tarefa
    ;   Escolha = 0 -> 
        true  % Voltar sem fazer alterações
    ;   writeln('Opção inválida, tente novamente.'),
        acessar_sprint(Usuario, Sprint)  % Tentar novamente em caso de erro
    ).

% Exemplo de implementação da função para imprimir uma tarefa
print_tarefa(TarefaId) :- 
    tarefa(TarefaId, Titulo, Descricao, Prioridade, Status, CriadorId, ResponsavelId, _),
    format('ID: ~w, Título: ~w, Descrição: ~w, Prioridade: ~w, Status: ~w, Criador: ~w, Responsável: ~w~n', 
           [TarefaId, Titulo, Descricao, Prioridade, Status, CriadorId, ResponsavelId]).

% Função para adicionar uma nova tarefa à sprint
adicionar_tarefa_a_sprint(Sprint) :- 
    write('Digite o ID da tarefa a ser adicionada à sprint:'), nl,
    read(TarefaId),
    (   tarefa(TarefaId, Titulo, Descricao, Prioridade, backlog, CriadorId, _, EmpresaId) ->  % Verifica se a tarefa existe
        retract(sprint(Sprint, Nome, Duracao, Tarefas, CriadorId, EmpresaId)),
        assertz(sprint(Sprint, Nome, Duracao, [TarefaId | Tarefas], CriadorId, EmpresaId)),
        writeln('Tarefa adicionada à sprint com sucesso!')
    ;   writeln('Tarefa não encontrada ou não está no status "Backlog".')
    ).

% Função para atribuir tarefa a um usuário
atribuir_tarefa(Usuario) :- 
    write('Digite o ID da tarefa que deseja atribuir:'), nl,
    read(TarefaId),
    usuario_empresa_id(Usuario, EmpresaId),  % Obtendo o EmpresaId do usuário

    (   tarefa(TarefaId, Titulo, Descricao, Prioridade, backlog, CriadorId, _, EmpresaId) ->  % Verifica se a tarefa existe e tem status 'backlog'
        tipo_usuario(Usuario, dev_team),  % Verifica se o usuário que está atribuindo a tarefa é do tipo 'dev_team'
        write('Digite o ID do usuário para atribuir a tarefa:'), nl,
        read(UsuarioId),

        % Verifica se o UsuarioId é do tipo dev_team
        tipo_usuario(UsuarioId, dev_team),
        
        % Atualiza a tarefa para que o responsável seja o UsuarioId
        retract(tarefa(TarefaId, Titulo, Descricao, Prioridade, backlog, CriadorId, _, EmpresaId)),
        assertz(tarefa(TarefaId, Titulo, Descricao, Prioridade, backlog, CriadorId, UsuarioId, EmpresaId)),  % Adiciona a tarefa atualizada
        writeln('Tarefa atribuída com sucesso!')
    ;   writeln('Tarefa não encontrada ou não está no status correto.')
    ).

% Função auxiliar para imprimir a sprint
print_sprint(sprint(Id, Nome, Duracao, _, CriadorId, EmpresaId)) :- 
    format('ID: ~w, Nome: ~w, Duração: ~w dias~n', [Id, Nome, Duracao]).

File: haskell/src/Sprint.hs
module Sprint where

import Data.List (find) -- Importa a função find do módulo Data.List
import Data.Maybe (fromMaybe)
import Tarefa (StatusTarefa (..), Tarefa (..))
import Usuario (Usuario (..),TipoUsuario(..))

data Sprint = Sprint
  { sprintId :: Int,
    sprintNome :: String,
    sprintDuracao :: Int,
    sprintTarefas :: [Int],
    sprintIdCriador :: Int,
    sprintEmpresaId :: Int
  }
  deriving (Show, Eq)

criarSprint :: Usuario -> IO Sprint
criarSprint usuario = do
  putStrLn "Digite o ID da Sprint:"
  id <- readLn
  putStrLn "Digite o Nome da Sprint:"
  nome <- getLine
  putStrLn "Digite a Duração da Sprint (em dias):"
  duracao <- readLn
  let empresaId = usuarioEmpresaId usuario -- Pega o ID da Empresa do usuário
  let sprint = Sprint id nome duracao [] 0 empresaId
  putStrLn $ "Sprint criada: " ++ show sprint
  return sprint

listarSprint :: Usuario -> [Sprint] -> [Tarefa] -> IO ()
listarSprint usuario sprints tarefas = do
  let sprintsCriadas = filter (\s -> sprintIdCriador s == usuarioId usuario) sprints

  let tarefasUsuario = filter (\t -> tarefaIdCriador t == usuarioId usuario || tarefaIdResponsavel t == usuarioId usuario) tarefas

  let idsSprintsUsuario = map tarefaId tarefasUsuario

  let sprintsEnvolvido = filter (\sprint -> any (`elem` sprintTarefas sprint) idsSprintsUsuario) sprints

  let todasSprints = sprintsCriadas ++ sprintsEnvolvido

  let sprintsUnicas = removeDuplicatas todasSprints

  if null sprintsUnicas
    then putStrLn "O usuário não possui sprints."
    else do
      putStrLn "Sprints do usuário:"
      mapM_ (putStrLn . sprintNome) sprintsUnicas

-- Função auxiliar para remover duplicatas da lista de sprints
removeDuplicatas :: [Sprint] -> [Sprint]
removeDuplicatas = foldr (\x acc -> if x `elem` acc then acc else x : acc) []

-- Função para listar sprints da empresa
listarSprintsDaEmpresa :: Usuario -> [Sprint] -> [Usuario] -> [Tarefa] -> IO ([Sprint], [Tarefa])
listarSprintsDaEmpresa usuario sprints usuarios tarefas = do
  let sprintsDaEmpresa = filter (\s -> sprintEmpresaId s == usuarioEmpresaId usuario) sprints
  putStrLn "Sprints da Empresa:"
  mapM_ (\s -> putStrLn $ "ID: " ++ show (sprintId s) ++ ", Nome: " ++ sprintNome s) sprintsDaEmpresa
  putStrLn "Digite o ID da sprint para visualizar suas tarefas (ou -1 para voltar, 0 para criar uma nova sprint):"
  entrada <- getLine
  case entrada of
    "-1" -> return (sprints, tarefas)
    "0" -> do
      novaSprint <- criarSprint usuario
      let sprintsAtualizados = novaSprint : sprints
      listarSprintsDaEmpresa usuario sprintsAtualizados usuarios tarefas
    _ -> do
      let sprintIdEscolhida = read entrada
      case find (\s -> sprintId s == sprintIdEscolhida) sprintsDaEmpresa of
        Nothing -> putStrLn "Sprint não encontrada." >> listarSprintsDaEmpresa usuario sprints usuarios tarefas
        Just sprint -> do
          (sprintsAtualizados, tarefasAtualizadas) <- acessarSprint usuario sprint sprints tarefas usuarios
          listarSprintsDaEmpresa usuario sprintsAtualizados usuarios tarefasAtualizadas

acessarSprint :: Usuario -> Sprint -> [Sprint] -> [Tarefa] ->[Usuario] -> IO ([Sprint], [Tarefa])
acessarSprint usuario sprint sprints tarefas usuarios = do
  putStrLn $ "Sprint Selecionada: " ++ sprintNome sprint
  putStrLn "Tarefas da Sprint:"
  let tarefasDaSprint = filter (\t -> tarefaId t `elem` sprintTarefas sprint) tarefas
  mapM_ (\t -> putStrLn $ "ID: " ++ show (tarefaId t) ++ ", Título: " ++ tarefaTitulo t ++ ", Status: " ++ show (tarefaStatus t)) tarefasDaSprint
  putStrLn "Escolha uma opção:"
  putStrLn "1. Adicionar Tarefa à Sprint"
  putStrLn "2. Atribuir Tarefa a um Usuário"
  putStrLn "0. Voltar"
  escolha <- getLine
  case escolha of
    "1" -> do
      (sprintAtualizada, sprintsAtualizados, tarefasAtualizadas) <- adicionarTarefaASprint usuario sprint sprints tarefas
      acessarSprint usuario sprintAtualizada sprintsAtualizados tarefasAtualizadas usuarios
    "2" -> do
      tarefasAtualizadas <- atribuirTarefa usuario tarefas usuarios
      acessarSprint usuario sprint sprints tarefasAtualizadas usuarios
    "0" -> return (sprints, tarefas)
    _   -> do
      putStrLn "Opção inválida, tente novamente."
      acessarSprint usuario sprint sprints tarefas usuarios
-- Função para adicionar uma tarefa à sprint
-- Função para adicionar uma tarefa à sprint
adicionarTarefaASprint :: Usuario -> Sprint -> [Sprint] -> [Tarefa] -> IO (Sprint, [Sprint], [Tarefa])
adicionarTarefaASprint usuario sprint sprints tarefas = do
  putStrLn "Digite o ID da tarefa para adicionar à sprint:"
  tarefaIdEscolhida <- readLn
  case find (\t -> tarefaId t == tarefaIdEscolhida && tarefaStatus t == Backlog) tarefas of
    Nothing -> do
      putStrLn "Tarefa não encontrada ou não está no status 'Backlog'."
      return (sprint, sprints, tarefas)
    Just tarefa -> do
      let tarefaAtualizada = tarefa { tarefaStatus = Pendente }
          tarefasAtualizadas = map (\t -> if tarefaId t == tarefaIdEscolhida then tarefaAtualizada else t) tarefas
          
          sprintAtualizada = sprint { sprintTarefas = tarefaIdEscolhida : sprintTarefas sprint }
          sprintsAtualizados = map (\s -> if sprintId s == sprintId sprint then sprintAtualizada else s) sprints
          
      putStrLn "Tarefa adicionada à sprint e status atualizado com sucesso!"
      return (sprintAtualizada, sprintsAtualizados, tarefasAtualizadas)
-- Função para atribuir uma tarefa a um usuário
atribuirTarefa :: Usuario -> [Tarefa] -> [Usuario] -> IO [Tarefa]
atribuirTarefa usuario tarefas usuarios = do
  putStrLn "Digite o ID da tarefa para atribuir a um usuário:"
  tarefaIdEscolhida <- readLn
  let tarefaSelecionada = find (\t -> tarefaId t == tarefaIdEscolhida && tarefaStatus t == Pendente) tarefas
  case tarefaSelecionada of
    Nothing -> do
      putStrLn "Tarefa não encontrada ou não está no status 'Pendente'."
      return tarefas -- Retorna as listas originais, pois a tarefa não foi encontrada
    Just tarefa -> do
      putStrLn "Digite o ID do usuário para atribuir a tarefa:"
      usuarioIdEscolhido <- readLn
      case find (\u -> usuarioId u == usuarioIdEscolhido && usuarioEmpresaId u == tarefaEmpresaId tarefa && usuarioPapel u == DevTeam) usuarios of
        Nothing -> do
          putStrLn "Usuário não encontrado, não pertence à empresa ou não é um desenvolvedor."
          return tarefas -- Retorna as listas originais, pois o usuário não é válido
        Just usuarioAtribuido -> do
          -- Atualiza a tarefa com o responsável e o status
          let tarefaAtualizada = tarefa { tarefaIdResponsavel =  usuarioId usuarioAtribuido, tarefaStatus = Pendente }
              tarefasAtualizadas = map (\t -> if tarefaId t == tarefaIdEscolhida then tarefaAtualizada else t) tarefas
          putStrLn "Tarefa atribuída com sucesso!"
          return tarefasAtualizadas